packages/babel-parser/src/tokenizer/state.ts

```diff
@@ -43,17 +44,28 @@ export default class State {
   startLoc: Position;
   endLoc: Position;
 
-  init({ strictMode, sourceType, startLine, startColumn }: Options): void {
+  init({
+    strictMode,
+    sourceType,
+    startIndex,
+    startLine,
+    startColumn,
+  }: Options): void {
     this.strict =
       strictMode === false
         ? false
         : strictMode === true
           ? true
           : sourceType === "module";
 
+    this.startIndex = startIndex;
     this.curLine = startLine;
     this.lineStart = -startColumn;
-    this.startLoc = this.endLoc = new Position(startLine, startColumn, 0);
+    this.startLoc = this.endLoc = new Position(
+      startLine,
+      startColumn,
+      startIndex,
```

Comment:
I think `startIndex` makes sense only when `startLine` > 1, otherwise we should have just set `startIndex` to the `startColumn`.

If one set `startIndex` without `startLine` and `startColumn`, the location info will still be incorrect. Can we throw if `startIndex` is provided without `startLine`?

Reply:
> I think startIndex makes sense only when startLine > 1
Technically it could be defaulted to startColumn. I personally think it's fine to say the the consumer of the parser (presumably building a tool on top of babel) knows to pass everything.

I think you could argue that it should error without passing `startIndex` regardless since even now the location information is incorrect (however this would be a breaking change). 

I don't know that it's a huge deal to error here though.

Reply:
FWIW In [marko-js/marko](https://github.com/marko-js/marko) we are using @babel/parser and have historically used a _hack_ to get correct source node location information by [prefixing all parsed code with the correct number of new lines and spaces](https://github.com/marko-js/marko/pull/2274/files#diff-a5f60e6e00e732d91e3c86273b501d356c3501a66f2e6e14cdb43cb0dbbe3a34L126-L127). However we saw that in really large templates it destroyed parsing performance to do this. Right now we're in a strange in between where we just assume the `index` information is incorrect and use a helper to go from loc to index.

Anyways just some context.

If there's anything I can do to help you all get this in I am happy to do so.

Reply:
I don't object the `startIndex` option, as it complements line and column to allow full start position customization. If the current incorrect index behaviour is depended, we can ship it in Babel 8 (wrap the default logic within the `process.env.BABEL_8_BREAKING` condition). I do believe a good default value is important: if one is using `startColumn`, they will get improved location indexes for free. You can always override the default value if you like.

We can add the default logic to the `getOptions` function so the tokenizer don't have to take care of it:
https://github.com/babel/babel/blob/4ff940865fe4e2ea6485e3712aeb4e32e7e8344d/packages/babel-parser/src/options.ts#L96-L102

Reply:
@JLHwung I'm not entirely following what the proposed code change is... please bear with me.

Are you suggesting something like

```
  if (
    options.startIndex > 0 &&
    options.startLine === 1 &&
    typeof opts.startColumn !== "number"
  ) {
    options.startColumn = options.startIndex;
  }
```

Reply:
FYI @JLHwung I've updated the PR with the above change!

Reply:
Actually I think I follow you now. I updated the code to error when `startIndex` is used along side a `startLine > 1` without a `startColumn`. Added a test for that also!

LMK if there's anything else I can do 😄

Reply:
It's closed, actually I meant the opposite. As `startColumn` was introduced before `startIndex`, it is expected that more users set `startColumn` than `startIndex`.

```js
if (
    options.startColumn > 0 &&
    options.startLine === 1 &&
    opts.startIndex == null
  ) {
    options.startIndex = options.startColumn;
  }
```